File: t/CS/OmrProcessingDemo/UI/OMRProcessInputWizard/ErrorStatisticsPanel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OmrProcessingDemo.UI
{
   class ErrorStatisticsPanel : WizardStep
   {
      private System.Windows.Forms.ListView lstResultReport;
      private Workspace.ReviewCounter answerReviewCounts;
      private Workspace.ReviewCounter resultsReviewCounts;
      public ErrorStatisticsPanel(Workspace.ReviewCounter answerReviewCounts, Workspace.ReviewCounter resultsReviewCounts) : base()
      {
         InitializeComponent();

         this.answerReviewCounts = answerReviewCounts;
         this.resultsReviewCounts = resultsReviewCounts;

         PopulateListBox();

         Title = "Report";
      }

      private void PopulateListBox()
      {
         lstResultReport.Columns.Add("Analysis: ");

         if (answerReviewCounts.IsEmpty == false)
         {
            lstResultReport.Columns.Add("Found in Answer Key");
         }
         if (resultsReviewCounts.IsEmpty == false)
         {
            lstResultReport.Columns.Add("Found in Worksheets");
         }

         PopulateValues("Below confidence threshold", answerReviewCounts.BelowThreshold, resultsReviewCounts.BelowThreshold);
         PopulateValues("Not exactly one bubble selected", answerReviewCounts.NotExactlyOneBubble, resultsReviewCounts.NotExactlyOneBubble);

         //if ((answerReviewCounts.IsEmpty && resultsReviewCounts.AgreeWithKey == 0 && resultsReviewCounts.DisagreeWithKey == 0) == false)
         //{
         //   PopulateValues("Same as key (Correct answer)", answerReviewCounts.AgreeWithKey, resultsReviewCounts.AgreeWithKey);
         //   PopulateValues("Different from key (Incorrect answer)", answerReviewCounts.DisagreeWithKey, resultsReviewCounts.DisagreeWithKey);
         //}

         if ((answerReviewCounts.IsEmpty && resultsReviewCounts.AgreeWithKey == 0 && resultsReviewCounts.DisagreeWithKey == 0) == false)
         {
            PopulateValues("Same as key (Correct answer)", "N/A", resultsReviewCounts.AgreeWithKey);
            PopulateValues("Different from key (Incorrect answer)", "N/A", resultsReviewCounts.DisagreeWithKey);
         }

         //PopulateValues("Totals", answerReviewCounts.Total, resultsReviewCounts.Total);
         lstResultReport.Items.Add("");
         PopulateValues("Total Collections", answerReviewCounts.TotalCounts, resultsReviewCounts.TotalCounts);

         lstResultReport.AutoResizeColumns(System.Windows.Forms.ColumnHeaderAutoResizeStyle.HeaderSize);
         lstResultReport.AutoResizeColumn(0, System.Windows.Forms.ColumnHeaderAutoResizeStyle.ColumnContent);
      }

      private void PopulateValues(string v, int answerCount, int resultCount)
      {
         List<string> items = new List<string>();

         items.Add(v);
         if (answerReviewCounts.IsEmpty == false)
         {
            items.Add(answerCount.ToString());
         }
         if (resultsReviewCounts.IsEmpty == false)
         {
            items.Add(resultCount.ToString());
         }

         System.Windows.Forms.ListViewItem lvi = new System.Windows.Forms.ListViewItem(items.ToArray());

         lstResultReport.Items.Add(lvi);
      }

      private void PopulateValues(string v, string answerString, int resultCount)
      {
         List<string> items = new List<string>() { v, answerString, resultCount.ToString() };
         System.Windows.Forms.ListViewItem lvi = new System.Windows.Forms.ListViewItem(items.ToArray());
         lstResultReport.Items.Add(lvi);
      }

      private void InitializeComponent()
      {
         this.lstResultReport = new System.Windows.Forms.ListView();
         this.SuspendLayout();
         // 
         // lstResultReport
         // 
         this.lstResultReport.Dock = System.Windows.Forms.DockStyle.Fill;
         this.lstResultReport.HeaderStyle = System.Windows.Forms.ColumnHeaderStyle.Nonclickable;
         this.lstResultReport.Location = new System.Drawing.Point(0, 0);
         this.lstResultReport.MultiSelect = false;
         this.lstResultReport.Name = "lstResultReport";
         this.lstResultReport.Size = new System.Drawing.Size(617, 198);
         this.lstResultReport.TabIndex = 0;
         this.lstResultReport.UseCompatibleStateImageBehavior = false;
         this.lstResultReport.View = System.Windows.Forms.View.Details;
         // 
         // ErrorStatisticsPanel
         // 
         this.Controls.Add(this.lstResultReport);
         this.Name = "ErrorStatisticsPanel";
         this.Size = new System.Drawing.Size(617, 198);
         this.ResumeLayout(false);

      }
   }
}

File: t/VB/OmrProcessingDemo/UI/OMRProcessInputWizard/ErrorStatisticsPanel.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Class ErrorStatisticsPanel
   Inherits WizardStep

   Private _lstResultReport As System.Windows.Forms.ListView
   Private _answerReviewCounts As Workspace.ReviewCounter
   Private _resultsReviewCounts As Workspace.ReviewCounter

   Public Sub New(ByVal answerReviewCounts As Workspace.ReviewCounter, ByVal resultsReviewCounts As Workspace.ReviewCounter)
      MyBase.New()
      InitializeComponent()
      _answerReviewCounts = answerReviewCounts
      _resultsReviewCounts = resultsReviewCounts
      PopulateListBox()
      Title = "Report"
   End Sub

   Private Sub PopulateListBox()
      _lstResultReport.Columns.Add("Analysis: ")

      If _answerReviewCounts.IsEmpty = False Then
         _lstResultReport.Columns.Add("Found in Answer Key")
      End If

      If _resultsReviewCounts.IsEmpty = False Then
         _lstResultReport.Columns.Add("Found in Worksheets")
      End If

      PopulateValues("Below confidence threshold", _answerReviewCounts.BelowThreshold, _resultsReviewCounts.BelowThreshold)
      PopulateValues("Not exactly one bubble selected", _answerReviewCounts.NotExactlyOneBubble, _resultsReviewCounts.NotExactlyOneBubble)

      If (_answerReviewCounts.IsEmpty AndAlso _resultsReviewCounts.AgreeWithKey = 0 AndAlso _resultsReviewCounts.DisagreeWithKey = 0) = False Then
         PopulateValues("Same as key (Correct answer)", "N/A", _resultsReviewCounts.AgreeWithKey)
         PopulateValues("Different from key (Incorrect answer)", "N/A", _resultsReviewCounts.DisagreeWithKey)
      End If

      _lstResultReport.Items.Add("")
      PopulateValues("Total Collections", _answerReviewCounts.TotalCounts, _resultsReviewCounts.TotalCounts)
      _lstResultReport.AutoResizeColumns(ColumnHeaderAutoResizeStyle.HeaderSize)
      _lstResultReport.AutoResizeColumn(0, ColumnHeaderAutoResizeStyle.ColumnContent)
   End Sub

   Private Sub PopulateValues(ByVal v As String, ByVal answerCount As Integer, ByVal resultCount As Integer)
      Dim items As List(Of String) = New List(Of String)()
      items.Add(v)

      If _answerReviewCounts.IsEmpty = False Then
         items.Add(answerCount.ToString())
      End If

      If _resultsReviewCounts.IsEmpty = False Then
         items.Add(resultCount.ToString())
      End If

      Dim lvi As System.Windows.Forms.ListViewItem = New System.Windows.Forms.ListViewItem(items.ToArray())
      _lstResultReport.Items.Add(lvi)
   End Sub

   Private Sub PopulateValues(ByVal v As String, ByVal answerString As String, ByVal resultCount As Integer)
      Dim items As List(Of String) = New List(Of String)() From {
          v,
          answerString,
          resultCount.ToString()
      }
      Dim lvi As System.Windows.Forms.ListViewItem = New System.Windows.Forms.ListViewItem(items.ToArray())
      _lstResultReport.Items.Add(lvi)
   End Sub

   Private Sub InitializeComponent()
      _lstResultReport = New System.Windows.Forms.ListView()
      Me.SuspendLayout()
      _lstResultReport.Dock = System.Windows.Forms.DockStyle.Fill
      _lstResultReport.HeaderStyle = System.Windows.Forms.ColumnHeaderStyle.Nonclickable
      _lstResultReport.Location = New System.Drawing.Point(0, 0)
      _lstResultReport.MultiSelect = False
      _lstResultReport.Name = "lstResultReport"
      _lstResultReport.Size = New System.Drawing.Size(617, 198)
      _lstResultReport.TabIndex = 0
      _lstResultReport.UseCompatibleStateImageBehavior = False
      _lstResultReport.View = System.Windows.Forms.View.Details
      Me.Controls.Add(_lstResultReport)
      Me.Name = "ErrorStatisticsPanel"
      Me.Size = New System.Drawing.Size(617, 198)
      Me.ResumeLayout(False)
   End Sub
End Class
